File: relay/smtp.ts
import nodemailer from "nodemailer";
import type { Transporter } from "nodemailer";

export interface SmtpConfig {
  host: string;
  port: number;
  secure?: boolean;
  user: string;
  pass: string;
  from: string;
}

let transporter: Transporter | null = null;
let activeConfig: SmtpConfig | null = null;

export function isSmtpConfigured(config: SmtpConfig | null | undefined): boolean {
  return Boolean(config?.host && config.user && config.from);
}

export function getSmtpTransporter(config: SmtpConfig): Transporter {
  if (
    transporter &&
    activeConfig &&
    activeConfig.host === config.host &&
    activeConfig.user === config.user
  ) {
    return transporter;
  }

  transporter = nodemailer.createTransport({
    host: config.host,
    port: config.port,
    secure: config.secure ?? config.port === 465,
    auth: {
      user: config.user,
      pass: config.pass,
    },
  });
  activeConfig = config;
  return transporter;
}

export interface SendMailInput {
  to: string;
  subject: string;
  text: string;
  html?: string;
}

export async function sendMail(
  config: SmtpConfig,
  input: SendMailInput,
): Promise<{ messageId: string }> {
  const transport = getSmtpTransporter(config);
  const info = await transport.sendMail({
    from: config.from,
    to: input.to,
    subject: input.subject,
    text: input.text,
    html: input.html,
  });
  return { messageId: info.messageId };
}

File: src-tauri/src/relay/smtp.rs
use lettre::{
    message::{header::ContentType, MultiPart, SinglePart},
    transport::smtp::authentication::Credentials,
    AsyncSmtpTransport, AsyncTransport, Message, Tokio1Executor,
};

use crate::relay::config::SmtpConfig;

fn build_transport(config: &SmtpConfig) -> Result<AsyncSmtpTransport<Tokio1Executor>, String> {
    let creds = Credentials::new(config.user.clone(), config.pass.clone());

    let builder = if config.secure {
        AsyncSmtpTransport::<Tokio1Executor>::relay(&config.host)
            .map_err(|e| e.to_string())?
    } else {
        AsyncSmtpTransport::<Tokio1Executor>::starttls_relay(&config.host)
            .map_err(|e| e.to_string())?
    };

    Ok(builder.port(config.port).credentials(creds).build())
}

pub async fn test_connection(config: &SmtpConfig) -> Result<(), String> {
    let transport = build_transport(config)?;
    transport
        .test_connection()
        .await
        .map_err(|e| e.to_string())?;
    Ok(())
}

pub async fn send_mail(
    config: &SmtpConfig,
    to: &str,
    subject: &str,
    text: &str,
    html: Option<&str>,
) -> Result<String, String> {
    let from = config
        .from
        .parse()
        .map_err(|e: lettre::address::AddressError| e.to_string())?;
    let to_addr = to
        .parse()
        .map_err(|e: lettre::address::AddressError| e.to_string())?;

    let email = if let Some(html_body) = html {
        Message::builder()
            .from(from)
            .to(to_addr)
            .subject(subject)
            .multipart(
                MultiPart::alternative()
                    .singlepart(
                        SinglePart::builder()
                            .header(ContentType::TEXT_PLAIN)
                            .body(text.to_string()),
                    )
                    .singlepart(
                        SinglePart::builder()
                            .header(ContentType::TEXT_HTML)
                            .body(html_body.to_string()),
                    ),
            )
            .map_err(|e| e.to_string())?
    } else {
        Message::builder()
            .from(from)
            .to(to_addr)
            .subject(subject)
            .body(text.to_string())
            .map_err(|e| e.to_string())?
    };

    let transport = build_transport(config)?;
    transport.send(email).await.map_err(|e| e.to_string())?;

    Ok(format!("<{}@hesia.local>", uuid::Uuid::new_v4()))
}
